src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -181,142 +187,184 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
-                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
-              partitionInputStream =
-                  new SequenceInputStream(beginningOfRecord, concatenateEndOfRecord);
-              parser =
-                  new CSVParser(
-                      new BufferedReader(new InputStreamReader(partitionInputStream)),
-                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
-                      bytePositionAtBeginningOfRecord,
-                      0);
-              iterator = parser.iterator();
-              onLastRow = true;
-              record = iterator.next();
+
+      if (record != null) {
+        /*
+         * Here, we read the first byte of the stream if this is not the first worker. This is simply an optimization.
+         * If the previous character is not a new line, we discard the line.
+         */
+        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
+          char currentChar = (char) partitionInputStream.read();
+          if (currentChar != '\n' && currentChar != '\r') {
+            discardedRecord = true;
+            /*
+             * This handles closes the stream in corner cases in which a worker might only have a fragment of a tuple as
+             * its entire partition. Since the server makes sure to provide workers a partition with a minimum size,
+             * this case could only happen with very large rows.
+             */
+            if (onLastRow) {
+              parser.close();
               break;
             }
-            dataChar = startStreamReader.read();
+          } else {
+            /*
+             * This handles the case in which the splitting occurs in the middle of \r\n.
+             */
+            currentChar = (char) partitionInputStream.read();
+            if (currentChar == '\n') {
+              discardedRecord = true;
+            } else {
+              partitionInputStream =
+                  source.getInputStream(adjustedStartByteRange, adjustedEndByteRange);
+            }
           }
-          startStreamReader.close();
         }
-      }
-      // This is a partial fragment, columns are missing and we need to read more bytes
-      if (record.size() < schema.numColumns()) {
-        if (lineNumber - 1 != 0 && !isLastWorker) {
-          onLastRow = true;
-          long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
-          if (!fixingStartByte) {
-            fixingStartByte = true;
-            partitionStartByteRange += bytePositionAtBeginningOfRecord;
+
+        /*
+         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
+         * we have read the entire file (again, this is for the case where a single worker might be reading a single
+         * large row that was split among other workers)
+         */
+        if (record.size() == schema.numColumns() && onLastRow && !isLastWorker) {
```

Comment:
I agree that assuming any `IOException` indicates a truncated field wouldn't be safe, and relying on the exception message is fragile. Looks like we're not the only ones who have run into this issue: http://stackoverflow.com/questions/33464248/why-are-classes-declared-as-final-in-apache-commons-csv-library. There's an open JIRA for this issue, but it doesn't look there's any interest in fixing it: https://issues.apache.org/jira/browse/CSV-141.

I think for now, unfortunately, the simplest thing to do would be to match on this exception message (with a FIXME comment), and continue to read the rest of the record. I don't know how we could patch this to provide generally acceptable behavior (for our purposes, the parser shouldn't just ignore these records, but for many scenarios it should). It's actually somewhat convenient for us that the parser throws an exception on malformed input like this instead of ignoring it, or we wouldn't know there was a trailing record at all!

Note that we need to test the converse scenario where the first field in a record has no opening quote, since we'll need to handle this in the partition that follows a truncated record. Could you test this scenario and record the outcome (presumably an exception) in the PR? We'll have to figure out what to do in that case; if we hit an exception we might need to scan to the first newline in the partition and restart the parser there. Actually, come to think about it, what if we did that in general: check the preceding byte of the range for a newline, and if no newline precedes the range, scan to the first newline in the `InputStream`, then instantiate the `Reader` and `CSVParser`? That might actually simplify the code a bit, since we'd never be reading a partial (and possibly malformed) first record unless it extended past the partition.